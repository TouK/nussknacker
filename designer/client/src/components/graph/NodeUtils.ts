/* eslint-disable i18next/no-literal-string */
import _, {uniqBy} from "lodash"
import * as ProcessDefinitionUtils from "../../common/ProcessDefinitionUtils"
import ProcessUtils from "../../common/ProcessUtils"
import {
  Edge,
  EdgeKind,
  EdgeType,
  NodeId,
  NodeType,
  Process,
  ProcessDefinitionData,
  PropertiesType,
  SubprocessNodeType,
  UINodeType,
} from "../../types"
import {UnknownRecord} from "../../types/common"

class NodeUtils {

  isNode = (obj: UnknownRecord): obj is NodeType => {
    return !_.isEmpty(obj) && _.has(obj, "id") && _.has(obj, "type")
  }

  nodeType = (node: NodeType) => {
    return node.type ? node.type : "Properties"
  }

  nodeIsProperties = (node: UINodeType): node is PropertiesType => {
    const type = node && this.nodeType(node)
    return type === "Properties"
  }

  nodeIsSubprocess = (node): node is SubprocessNodeType => {
    return this.nodeType(node) === "SubprocessInput"
  }

  isPlainNode = (node: UINodeType) => {
    return !_.isEmpty(node) && !this.nodeIsProperties(node)
  }

  nodeIsJoin = (node: NodeType): boolean => {
    return node && this.nodeType(node) === "Join"
  }

  nodesFromProcess = (process: Process): NodeType[] => process.nodes || []

  edgesFromProcess = (process: Process) => process.edges || []

  getProcessProperties = ({id, properties}: Process, name?: string) => ({id: name || id, ...properties})

  getNodeById = (nodeId: NodeId, process: Process) => this.nodesFromProcess(process).find(n => n.id === nodeId)

  getEdgeById = (edgeId: NodeId, process: Process) => this.edgesFromProcess(process).find(e => this.edgeId(e) === edgeId)

  getAllNodesById = (nodeIds: NodeId[], process: Process) => {
    const allNodes = this.nodesFromProcess(process).filter(node => nodeIds.includes(node.id))
    return uniqBy(allNodes, n => n.id)
  }

  getAllNodesByIdWithEdges = (ids: NodeId[], process: Process) => {
    const nodes = this.getAllNodesById(ids, process)
    const edgesForNodes = this.getEdgesForConnectedNodes(nodes.map(n => n.id), process)
    return {
      nodes: nodes,
      edges: edgesForNodes,
    }
  }

  isAvailable = (node: NodeType, processDefinitionData, category): boolean => {
    const availableIdsInCategory = ProcessDefinitionUtils.getFlatCategoryComponents(processDefinitionData, category)
      .map(component => ProcessUtils.findNodeDefinitionIdOrType(component.node))
    const nodeDefinitionId = ProcessUtils.findNodeDefinitionIdOrType(node)
    return availableIdsInCategory.includes(nodeDefinitionId)
  }

  getIncomingEdges = (nodeId: NodeId, process: Process): Edge[] => this.edgesFromProcess(process).filter(e => e.to === nodeId)

  getEdgesForConnectedNodes = (
    nodeIds: NodeId[],
    process: Process,
  ): Edge[] => process.edges?.filter(edge => nodeIds.includes(edge.from) && nodeIds.includes(edge.to))

  edgeType = (allEdges: Edge[], node: NodeType, processDefinitionData: ProcessDefinitionData): EdgeType => {
    const edgesForNode = this.edgesForNode(node, processDefinitionData)

    if (edgesForNode.canChooseNodes) {
      return edgesForNode.edges[0]
    } else {
      const currentConnectionsTypes = allEdges.filter((edge) => edge.from === node.id).map(e => e.edgeType)
      return edgesForNode.edges.find(et => !currentConnectionsTypes.find(currentType => _.isEqual(currentType, et)))
    }
  }

  edgesForNode = (node: NodeType, processDefinitionData: ProcessDefinitionData, forInput?: boolean) => {
    const nodeObjectTypeDefinition = ProcessUtils.findNodeDefinitionId(node)
    //TODO: when we add more configuration for joins, probably more complex logic will be needed
    const data = processDefinitionData.edgesForNodes
      .filter(e => !forInput || e.isForInputDefinition === forInput)
      //here we use == in second comparison, as we sometimes compare null to undefined :|
      .find(e => e.nodeId.type === _.get(node, "type") && e.nodeId.id == nodeObjectTypeDefinition)
    return data || {edges: [null], canChooseNodes: false}
  }

  edgeLabel = (edge: Edge) => {
    const edgeType = edge?.edgeType
    const type = edgeType?.type
    switch (type) {
      case EdgeKind.subprocessOutput:
        return edgeType?.name
      case EdgeKind.switchNext:
        return edgeType?.condition?.expression
    }
    return this.edgeTypeLabel(type)
  }

  //TODO: i18next
  edgeTypeLabel = (type: string) => {
    switch (type) {
      case EdgeKind.filterFalse:
        return "ðŸ”´ false"
      case EdgeKind.filterTrue:
        return "ðŸŸ¢ true"
      case EdgeKind.switchDefault:
        return "default (deprecated)"
      case EdgeKind.switchNext:
        return "condition"
      default:
        return ""
    }
  }

  //we don't allow multi outputs other than split, filter, switch and no multiple inputs
  //TODO remove type (Source, Sink) comparisons
  canMakeLink = (fromId: NodeId, toId: NodeId, process: Process, processDefinitionData: ProcessDefinitionData, previousEdge?: Edge): boolean => {
    const nodeInputs = this._nodeInputs(toId, process)
    //we do not want to include currently edited edge
    const nodeOutputs = this._nodeOutputs(fromId, process)
      .filter(e => e.from !== previousEdge?.from && e.to !== previousEdge?.to)

    const to = this.getNodeById(toId, process)
    const from = this.getNodeById(fromId, process)
    if (fromId !== toId) {
      const canHaveMoreInputs = this.canHaveMoreInputs(to, nodeInputs, processDefinitionData)
      const canHaveMoreOutputs = this._canHaveMoreOutputs(from, nodeOutputs, processDefinitionData)
      const isUniqe = !nodeInputs.find(e => e.from === fromId)
      return canHaveMoreInputs && canHaveMoreOutputs && isUniqe
    }

    return false
  }

  canHaveMoreInputs = (node: NodeType, nodeInputs: Edge[], processDefinitionData: ProcessDefinitionData): boolean => {
    const edgesForNode = this.edgesForNode(node, processDefinitionData, true)
    const maxEdgesForNode = edgesForNode.edges.length
    return this.hasInputs(node) && (edgesForNode.canChooseNodes || nodeInputs.length < maxEdgesForNode)
  }

  _canHaveMoreOutputs = (node, nodeOutputs, processDefinitionData): boolean => {
    const edgesForNode = this.edgesForNode(node, processDefinitionData, false)
    const maxEdgesForNode = edgesForNode.edges.length
    return this.hasOutputs(node) && (edgesForNode.canChooseNodes || nodeOutputs.filter(e => e.to).length < maxEdgesForNode)
  }

  _nodeInputs = (nodeId: NodeId, process: Process) => {
    return this.edgesFromProcess(process).filter(e => e.to == nodeId)
  }

  _nodeOutputs = (nodeId: NodeId, process: Process) => {
    return this.edgesFromProcess(process).filter(e => e.from == nodeId)
  }

  edgeId = (edge: Edge): string => {
    return `${edge.from}-${edge.to}`
  }

  noInputNodeTypes = ["Source", "SubprocessInputDefinition"]

  noOutputNodeTypes = ["Sink", "SubprocessOutputDefinition"]

  //TODO: methods below should be based on backend data, e.g. Subprocess can have outputs or not - based on individual subprocess...
  hasInputs = (node: NodeType) => !this.noInputNodeTypes.some((nodeType) => _.isEqual(nodeType, node?.type))

  hasOutputs = (node: NodeType) => !this.noOutputNodeTypes.some((nodeType) => _.isEqual(nodeType, node?.type))

}

//TODO this pattern is not necessary, just export every public function as in actions.js
export default new NodeUtils()
