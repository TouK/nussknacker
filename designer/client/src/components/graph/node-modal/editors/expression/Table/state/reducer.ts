import { TableData } from "./tableState";
import { expandTable, longestRow } from "./helpers";
import { Action, ActionTypes } from "./action";

export function reducer(state: TableData, action: Action): TableData {
    switch (action.type) {
        case ActionTypes.insertData: {
            const newRowsNeeded = action.row + action.input.length - state.rows.length;
            const newColsNeeded = action.column + longestRow(action.input).length - state.columns.length;
            const updatedData = expandTable(state, newRowsNeeded, newColsNeeded, action.dataType);

            const rows = updatedData.rows.slice();
            action.input.forEach((columns, y) => {
                columns.forEach((value, x) => {
                    const currentRow = action.row + y;
                    if (currentRow >= 0) {
                        const row = rows[currentRow].slice();
                        row[action.column + x] = value;
                        rows[currentRow] = row;
                    }
                });
            });

            return {
                ...updatedData,
                rows,
            };
        }
        case ActionTypes.editData:
            return {
                ...state,
                rows: action.dataChanges.reduce(
                    (rows, { row: y, column: x, value }) => rows.map((r, i) => (i === y ? r.map((v, i) => (i === x ? value : v)) : r)),
                    state.rows,
                ),
            };
        case ActionTypes.renameColumn:
            // prevent duplicates
            if (state.columns.find(({ name }) => name === action.to)) {
                return state;
            }
            return {
                ...state,
                columns: state.columns.map(({ name, ...col }) => ({ name: name === action.from ? action.to : name, ...col })),
            };
        case ActionTypes.deleteColumns:
            return {
                ...state,
                columns: state.columns.filter((_, i) => !action.columns.includes(i)),
                rows: state.rows.map((r) => r.filter((_, i) => !action.columns.includes(i))),
            };
        case ActionTypes.deleteRows:
            return {
                ...state,
                rows: state.rows.filter((_, i) => !action.rows.includes(i)),
            };
        case ActionTypes.changeColumnType:
            if (!(action.column >= 0 && action.dataType)) {
                return state;
            }
            return {
                ...state,
                columns: state.columns.map((column, i) => {
                    if (i !== action.column) {
                        return column;
                    }
                    return {
                        ...column,
                        type: action.dataType,
                    };
                }),
            };
        case ActionTypes.columnResize:
            return {
                ...state,
                columns: state.columns.map?.((column, i) => {
                    if (i !== action.column) {
                        return column;
                    }
                    return {
                        ...column,
                        size: action.size.toString(),
                    };
                }),
            };
        case ActionTypes.resetColumnsSize:
            return {
                ...state,
                columns: state.columns.map(({ size, ...c }, i) =>
                    action.columns.includes(i)
                        ? c
                        : {
                              ...c,
                              size,
                          },
                ),
            };
        case ActionTypes.expand:
            return expandTable(state, action.rows, action.columns, action.dataType);
        case ActionTypes.replaceData:
            return action.data;
        default:
            return state;
    }
}
