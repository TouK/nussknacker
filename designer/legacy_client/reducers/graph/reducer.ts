/* eslint-disable i18next/no-literal-string */
import {isEqual, pick, sortBy} from "lodash"
import undoable, {combineFilters, excludeAction} from "redux-undo"
import {Reducer} from "../../actions/reduxTypes"
import {GraphState} from "./types"
import {batchGroupBy} from "./batchGroupBy"

//TODO: We should change namespace from graphReducer to currentlyDisplayedProcess

const emptyGraphState: GraphState = {
  graphLoading: false,
  processToDisplay: null,
  fetchedProcessDetails: null,
  testCapabilities: {},
  selectionState: [],
  processCounts: {},
  testResults: null,
  unsavedNewName: null,
}

const graphReducer: Reducer<GraphState> = (state = emptyGraphState, action) => {
  switch (action.type) {
    case "PROCESS_FETCH":
    case "PROCESS_LOADING": {
      return {
        ...state,
        graphLoading: true,
      }
    }
    case "UPDATE_TEST_CAPABILITIES": {
      return {
        ...state,
        testCapabilities: action.capabilities,
      }
    }
    case "LOADING_FAILED": {
      return {
        ...state,
        graphLoading: false,
      }
    }
    case "CLEAR_PROCESS": {
      return {
        ...state,
        processToDisplay: null,
        fetchedProcessDetails: null,
        testResults: null,
      }
    }
    case "URL_CHANGED": {
      return {
        ...state,
        ...emptyGraphState,
      }
    }
    case "HIDE_RUN_PROCESS_DETAILS": {
      return {
        ...state,
        testResults: null,
        processCounts: null,
      }
    }
    default:
      return state
  }
}

const undoableReducer = undoable(graphReducer, {
  ignoreInitialState: true,
  undoType: "UNDO",
  redoType: "REDO",
  clearHistoryType: ["CLEAR", "PROCESS_FETCH"],
  groupBy: batchGroupBy.init(),
  filter: combineFilters(
    excludeAction([
      "USER_TRACKING",
      "VALIDATION_RESULT",
      "DISPLAY_PROCESS",
      "UPDATE_IMPORTED_PROCESS",
      "PROCESS_STATE_LOADED",
      "UPDATE_BACKEND_NOTIFICATIONS",
    ]),
    (action, nextState, prevState) => {
      const keys: Array<keyof GraphState> = [
        "fetchedProcessDetails",
        "processToDisplay",
        "unsavedNewName",
        "selectionState",
      ]
      return !isEqual(
        pick(nextState, keys),
        pick(prevState._latestUnfiltered, keys),
      )
    },
  ),
})

//TODO: replace this with use of selectors everywhere
export function reducerWithUndo(state, action) {
  const history = undoableReducer(state?.history, action)
  return {...history.present, history}
}
