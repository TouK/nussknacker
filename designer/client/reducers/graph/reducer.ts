/* eslint-disable i18next/no-literal-string */
import {concat, isEqual, pick, sortBy, uniq, xor, zipObject} from "lodash"
import undoable, {combineFilters, excludeAction} from "redux-undo"
import {Reducer} from "../../actions/reduxTypes"
import * as GraphUtils from "../../components/graph/GraphUtils"
import * as LayoutUtils from "../layoutUtils"
import {nodes} from "../layoutUtils"
import {mergeReducers} from "../mergeReducers"
import {GraphState} from "./types"
import {
  addNodesWithLayout,
  adjustBranchParametersAfterDisconnect,
  createEdge,
  enrichNodeWithProcessDependentData,
  prepareNewNodesWithLayout,
  updateAfterNodeDelete,
  updateAfterNodeIdChange,
} from "./utils"
import {Edge, ValidationResult} from "../../types"
import NodeUtils from "../../components/graph/NodeUtils"
import {batchGroupBy} from "./batchGroupBy"

//TODO: We should change namespace from graphReducer to currentlyDisplayedProcess

const emptyGraphState: GraphState = {
  graphLoading: false,
  processToDisplay: null,
  fetchedProcessDetails: null,
  layout: [],
  testCapabilities: {},
  selectionState: [],
  processCounts: {},
  testResults: null,
  unsavedNewName: null,
}

export function updateValidationResult(state: GraphState, action: { validationResult: ValidationResult }): ValidationResult {
  return {
    ...action.validationResult,
    // nodeResults is sometimes empty although it shouldn't e.g. when SaveNotAllowed errors happen
    nodeResults: {
      ...state.processToDisplay.validationResult.nodeResults,
      ...action.validationResult.nodeResults,
    },
  }
}

const graphReducer: Reducer<GraphState> = (state = emptyGraphState, action) => {
  switch (action.type) {
    case "PROCESS_FETCH":
    case "PROCESS_LOADING": {
      return {
        ...state,
        graphLoading: true,
      }
    }
    case "UPDATE_IMPORTED_PROCESS": {
      const oldNodeIds = sortBy(state.processToDisplay.nodes.map((n) => n.id))
      const newNodeids = sortBy(action.processJson.nodes.map((n) => n.id))
      const newLayout = isEqual(oldNodeIds, newNodeids) ? state.layout : null

      return {
        ...state,
        graphLoading: false,
        processToDisplay: action.processJson,
        layout: newLayout,
      }
    }
    case "UPDATE_TEST_CAPABILITIES": {
      return {
        ...state,
        testCapabilities: action.capabilities,
      }
    }
    case "DISPLAY_PROCESS": {
      const {fetchedProcessDetails} = action
      const processToDisplay = fetchedProcessDetails.json
      return {
        ...state,
        processToDisplay,
        fetchedProcessDetails,
        graphLoading: false,
        layout: LayoutUtils.fromMeta(processToDisplay),
      }
    }
    case "LOADING_FAILED": {
      return {
        ...state,
        graphLoading: false,
      }
    }
    case "CLEAR_PROCESS": {
      return {
        ...state,
        processToDisplay: null,
        fetchedProcessDetails: null,
        testResults: null,
      }
    }
    case "EDIT_NODE": {
      const stateAfterNodeRename = {
        ...state,
        ...updateAfterNodeIdChange(state.layout, action.processAfterChange, action.before.id, action.after.id),
      }
      return {
        ...stateAfterNodeRename,
        processToDisplay: {
          ...stateAfterNodeRename.processToDisplay,
          validationResult: updateValidationResult(state, action),
        },
      }
    }
    case "PROCESS_RENAME": {
      return {
        ...state,
        unsavedNewName: action.name,
      }
    }
    case "DELETE_NODES": {
      return action.ids.reduce((state, idToDelete) => {
        const stateAfterNodeDelete = updateAfterNodeDelete(state, idToDelete)
        const processToDisplay = GraphUtils.deleteNode(stateAfterNodeDelete.processToDisplay, idToDelete)
        return {
          ...stateAfterNodeDelete,
          processToDisplay,
        }
      }, state)
    }
    case "URL_CHANGED": {
      return {
        ...state,
        ...emptyGraphState,
      }
    }
    case "NODES_CONNECTED": {
      let newEdges: Edge[]

      const availableEdges = NodeUtils.edgesForNode(action.fromNode, action.processDefinitionData).edges
      const freeOutputEdges = state.processToDisplay.edges
        .filter(e => e.from === action.fromNode.id && !e.to)
        //we do this to skip e.g. edges that became incorrect/unavailable
        .filter(e => availableEdges.find(available => available?.name == e?.edgeType?.name && available?.type == e?.edgeType?.type))

      const freeOutputEdge = freeOutputEdges.find(e => e.edgeType === action.edgeType) || (freeOutputEdges.length == 0 ? null : freeOutputEdges[0])
      if (freeOutputEdge) {
        newEdges = state.processToDisplay.edges.map(e => e === freeOutputEdge ?
          {
            ...freeOutputEdge,
            to: action.toNode.id,
          } :
          e)
      } else {
        const edge = createEdge(action.fromNode, action.toNode, action.edgeType, state.processToDisplay.edges, action.processDefinitionData)
        newEdges = concat(state.processToDisplay.edges, edge)
      }

      return {
        ...state,
        processToDisplay: {
          ...state.processToDisplay,
          nodes: state.processToDisplay.nodes.map(n => action.toNode.id !== n.id ?
            n :
            enrichNodeWithProcessDependentData(
              n,
              action.processDefinitionData,
              newEdges,
            )),
          edges: newEdges,
        },
      }
    }
    case "NODES_DISCONNECTED": {
      const nodesToSet = adjustBranchParametersAfterDisconnect(state.processToDisplay.nodes, [action])
      return {
        ...state,
        processToDisplay: {
          ...state.processToDisplay,
          edges: state.processToDisplay.edges
            .map((e) => e.from === action.from && e.to === action.to ? {...e, to: ""} : e)
            .filter(Boolean),
          nodes: nodesToSet,
        },
      }
    }
    case "NODE_ADDED": {
      const nodeWithPosition = {
        node: action.node,
        position: action.position,
      }
      const {uniqueIds, nodes, layout} = prepareNewNodesWithLayout(state, [nodeWithPosition], false)
      return {
        ...addNodesWithLayout(state, {nodes, layout}),
        selectionState: uniqueIds,
      }
    }
    case "NODES_WITH_EDGES_ADDED": {
      const {nodes, layout, uniqueIds} = prepareNewNodesWithLayout(state, action.nodesWithPositions, true)

      const idToUniqueId = zipObject(action.nodesWithPositions.map(n => n.node.id), uniqueIds)
      const edgesWithValidIds = action.edges.map(edge => ({
        ...edge,
        from: idToUniqueId[edge.from],
        to: idToUniqueId[edge.to],
      }))

      const updatedEdges = edgesWithValidIds.reduce((edges, edge) => {
        const fromNode = nodes.find(n => n.id === edge.from)
        const toNode = nodes.find(n => n.id === edge.to)
        const newEdge = createEdge(fromNode, toNode, edge.edgeType, edges, action.processDefinitionData)
        return edges.concat(newEdge)
      }, state.processToDisplay.edges)

      const stateWithNodesAdded = addNodesWithLayout(state, {nodes, layout})
      return {
        ...stateWithNodesAdded,
        processToDisplay: {
          ...stateWithNodesAdded.processToDisplay,
          edges: updatedEdges,
        },
        selectionState: uniqueIds,
      }
    }
    case "VALIDATION_RESULT": {
      return {
        ...state,
        processToDisplay: {
          ...state.processToDisplay,
          validationResult: updateValidationResult(state, action),
        },
      }
    }
    //TODO: handle it differently?
    case "LAYOUT_CHANGED": {
      return {
        ...state,
        layout: action.layout,
      }
    }
    case "DISPLAY_PROCESS_COUNTS": {
      return {
        ...state,
        processCounts: action.processCounts,
      }
    }
    case "DISPLAY_TEST_RESULTS_DETAILS": {
      return {
        ...state,
        testResults: action.testResults,
        graphLoading: false,
      }
    }
    case "HIDE_RUN_PROCESS_DETAILS": {
      return {
        ...state,
        testResults: null,
        processCounts: null,
      }
    }
    case "EXPAND_SELECTION": {
      return {
        ...state,
        selectionState: uniq([...state.selectionState, ...action.nodeIds]),
      }
    }
    case "TOGGLE_SELECTION": {
      return {
        ...state,
        selectionState: xor(state.selectionState, action.nodeIds),
      }
    }
    case "RESET_SELECTION": {
      return {
        ...state,
        selectionState: action.nodeIds ? action.nodeIds : [],
      }
    }
    default:
      return state
  }
}

const reducer = mergeReducers(
  graphReducer,
  {
    processToDisplay: {
      nodes,
    },
  },
)

const undoableReducer = undoable(reducer, {
  ignoreInitialState: true,
  undoType: "UNDO",
  redoType: "REDO",
  clearHistoryType: ["CLEAR", "PROCESS_FETCH"],
  groupBy: batchGroupBy.init(),
  filter: combineFilters(
    excludeAction([
      "USER_TRACKING",
      "VALIDATION_RESULT",
      "DISPLAY_PROCESS",
      "UPDATE_IMPORTED_PROCESS",
      "PROCESS_STATE_LOADED",
      "UPDATE_BACKEND_NOTIFICATIONS",
    ]),
    (action, nextState, prevState) => {
      const keys: Array<keyof GraphState> = [
        "fetchedProcessDetails",
        "processToDisplay",
        "unsavedNewName",
        "layout",
        "selectionState",
      ]
      return !isEqual(
        pick(nextState, keys),
        pick(prevState._latestUnfiltered, keys),
      )
    },
  ),
})

//TODO: replace this with use of selectors everywhere
export function reducerWithUndo(state, action) {
  const history = undoableReducer(state?.history, action)
  return {...history.present, history}
}
