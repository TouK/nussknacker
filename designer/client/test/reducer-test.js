import reducer from "../src/reducers/index";
import NodeUtils from "../src/components/graph/NodeUtils";

const baseProcessState = {
    name: "DEFGH",
    isLatestVersion: true,
    processType: "graph",
    processCategory: "Category1",
    modificationDate: "2017-02-14T11:16:56.686",
    createdAt: "2017-02-14T11:16:56.686",
    createdBy: "admin",
    tags: [],
    currentlyDeployedAt: ["test"],
    scenarioGraph: {
        properties: {
            parallelism: 3,
            additionalFields: {
                groups: [
                    {
                        id: "acdc",
                        nodes: ["kafka-transaction", "filter"],
                    },
                ],
            },
        },
        nodes: [
            {
                type: "Source",
                id: "kafka-transaction",
                ref: {
                    typ: "kafka-transaction",
                    parameters: [],
                },
                additionalFields: {
                    description: "asdfasdfłóóódźźźasdfsdfasdfasdfasdfasdf",
                },
            },
            {
                type: "Filter",
                id: "filter",
                expression: {
                    language: "spel",
                    expression: "4 / (#input.length -5) >= 0",
                },
            },
            {
                type: "Enricher",
                id: "paramService",
                service: {
                    id: "paramService",
                    parameters: [
                        {
                            name: "param",
                            expression: {
                                language: "spel",
                                expression: "'3434'",
                            },
                        },
                    ],
                },
                output: "output",
            },
            {
                type: "Sink",
                id: "sendSms",
                ref: {
                    typ: "sendSms",
                    parameters: [],
                },
            },
        ],
        edges: [
            {
                from: "kafka-transaction",
                to: "filter",
            },
            {
                from: "filter",
                to: "paramService",
                edgeType: {
                    type: "FilterTrue",
                },
            },
            {
                from: "paramService",
                to: "sendSms",
            },
        ],
    },
    validationResult: {
        errors: {
            invalidNodes: {},
            processPropertiesErrors: [],
            globalErrors: [],
        },
    },
    state: {},
};

const baseState = reducer(
    {},
    {
        type: "@@INIT",
    },
);

const baseStateWithProcess = reducer(baseState, {
    type: "DISPLAY_PROCESS",
    scenario: baseProcessState,
});

const reduceAll = (actions) => actions.reduce((state, action) => reducer(state, action), baseStateWithProcess);

describe("Reducer suite", () => {
    it("Display process", () => {
        expect(baseStateWithProcess.graphReducer.scenario.name).toEqual(baseProcessState.name);
    });
});

const testNode = {
    type: "Enricher",
    id: "Enricher ID",
    service: {
        id: "paramService",
        parameters: [
            {
                name: "param",
                expression: {
                    language: "spel",
                    expression: "'3434'",
                },
            },
        ],
    },
    output: "output",
};

const testPosition = { x: 10, y: 20 };

describe("Nodes added", () => {
    it("should add single node", () => {
        const result = reduceAll([
            {
                type: "NODE_ADDED",
                node: testNode,
                position: testPosition,
            },
        ]);

        expect(NodeUtils.getNodeById(testNode.id, result.graphReducer.scenario.scenarioGraph)).toEqual(testNode);
        expect(result.graphReducer.layout.find((n) => n.id === testNode.id).position).toEqual(testPosition);
    });

    it("should add single node with unique id", () => {
        const result = reduceAll([
            {
                type: "NODE_ADDED",
                node: { ...testNode, id: "kafka-transaction" },
                position: testPosition,
            },
        ]);

        expect(NodeUtils.getNodeById("kafka-transaction 1", result.graphReducer.scenario.scenarioGraph)).toEqual({
            ...testNode,
            id: "kafka-transaction 1",
        });
        expect(result.graphReducer.layout.find((n) => n.id).position).toEqual(testPosition);
    });

    it("should add multiple nodes with unique id", () => {
        const result = reduceAll([
            {
                type: "NODES_WITH_EDGES_ADDED",
                nodesWithPositions: [
                    {
                        node: { ...testNode, id: "kafka-transaction" },
                        position: testPosition,
                    },
                    {
                        node: { ...testNode, id: "kafka-transaction" },
                        position: testPosition,
                    },
                ],
                edges: [],
            },
        ]);

        expect(NodeUtils.getNodeById("kafka-transaction (copy 1)", result.graphReducer.scenario.scenarioGraph)).toEqual({
            ...testNode,
            id: "kafka-transaction (copy 1)",
        });
        expect(NodeUtils.getNodeById("kafka-transaction (copy 2)", result.graphReducer.scenario.scenarioGraph)).toEqual({
            ...testNode,
            id: "kafka-transaction (copy 2)",
        });
    });

    it("should add nodes with edges", () => {
        const result = reduceAll([
            {
                type: "NODES_WITH_EDGES_ADDED",
                nodesWithPositions: [
                    {
                        node: { ...testNode, id: "newNode" },
                        position: testPosition,
                    },
                    {
                        node: { ...testNode, id: "kafka-transaction" },
                        position: testPosition,
                    },
                ],
                edges: [{ from: "newNode", to: "kafka-transaction" }],
                processDefinitionData: {
                    edgesForNodes: [],
                },
            },
        ]);

        expect(NodeUtils.getNodeById("newNode", result.graphReducer.scenario.scenarioGraph)).toEqual({ ...testNode, id: "newNode" });
        expect(NodeUtils.getNodeById("kafka-transaction (copy 1)", result.graphReducer.scenario.scenarioGraph)).toEqual({
            ...testNode,
            id: "kafka-transaction (copy 1)",
        });
        expect(NodeUtils.getEdgeById("newNode-kafka-transaction (copy 1)", result.graphReducer.scenario.scenarioGraph)).toEqual({
            from: "newNode",
            to: "kafka-transaction (copy 1)",
        });
    });
});
