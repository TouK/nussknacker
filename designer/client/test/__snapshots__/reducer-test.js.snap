// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nodes added should add multiple nodes with unique id 1`] = `
{
  "additionalFields": {
    "layoutData": {
      "x": 10,
      "y": 20,
    },
  },
  "branchParameters": undefined,
  "id": "kafka-transaction (copy 1)",
  "output": "output",
  "service": {
    "id": "paramService",
    "parameters": [
      {
        "expression": {
          "expression": "'3434'",
          "language": "spel",
        },
        "name": "param",
      },
    ],
  },
  "type": "Enricher",
}
`;

exports[`Nodes added should add multiple nodes with unique id 2`] = `
{
  "additionalFields": {
    "layoutData": {
      "x": 10,
      "y": 20,
    },
  },
  "branchParameters": undefined,
  "id": "kafka-transaction (copy 2)",
  "output": "output",
  "service": {
    "id": "paramService",
    "parameters": [
      {
        "expression": {
          "expression": "'3434'",
          "language": "spel",
        },
        "name": "param",
      },
    ],
  },
  "type": "Enricher",
}
`;

exports[`Nodes added should add nodes with edges 1`] = `
{
  "additionalFields": {
    "layoutData": {
      "x": 10,
      "y": 20,
    },
  },
  "branchParameters": undefined,
  "id": "newNode",
  "output": "output",
  "service": {
    "id": "paramService",
    "parameters": [
      {
        "expression": {
          "expression": "'3434'",
          "language": "spel",
        },
        "name": "param",
      },
    ],
  },
  "type": "Enricher",
}
`;

exports[`Nodes added should add nodes with edges 2`] = `
{
  "additionalFields": {
    "layoutData": {
      "x": 10,
      "y": 20,
    },
  },
  "branchParameters": undefined,
  "id": "kafka-transaction (copy 1)",
  "output": "output",
  "service": {
    "id": "paramService",
    "parameters": [
      {
        "expression": {
          "expression": "'3434'",
          "language": "spel",
        },
        "name": "param",
      },
    ],
  },
  "type": "Enricher",
}
`;

exports[`Nodes added should add single node 1`] = `
{
  "additionalFields": {
    "layoutData": {
      "x": 10,
      "y": 20,
    },
  },
  "branchParameters": undefined,
  "id": "Enricher ID",
  "output": "output",
  "service": {
    "id": "paramService",
    "parameters": [
      {
        "expression": {
          "expression": "'3434'",
          "language": "spel",
        },
        "name": "param",
      },
    ],
  },
  "type": "Enricher",
}
`;

exports[`Nodes added should add single node with unique id 1`] = `
{
  "additionalFields": {
    "layoutData": {
      "x": 10,
      "y": 20,
    },
  },
  "branchParameters": undefined,
  "id": "kafka-transaction 1",
  "output": "output",
  "service": {
    "id": "paramService",
    "parameters": [
      {
        "expression": {
          "expression": "'3434'",
          "language": "spel",
        },
        "name": "param",
      },
    ],
  },
  "type": "Enricher",
}
`;
